Fix bulk delete api, response code 400 - Bulk DELETE API is broken

praveenkjvs opened the pull request:
Changes to be committed:
	modified:   ambari-server/src/main/java/org/apache/ambari/server/api/handlers/DeleteHandler.java
	modified:   ambari-server/src/main/java/org/apache/ambari/server/controller/AmbariManagementControllerImpl.java
	modified:   ambari-server/src/main/java/org/apache/ambari/server/controller/internal/DeleteHostComponentStatusMetaData.java
	modified:   ambari-server/src/main/java/org/apache/ambari/server/state/ServiceComponentImpl.java

## What changes were proposed in this pull request?

https://issues.apache.org/jira/browse/AMBARI-25449

## How was this patch tested?

Manual

(If this patch involves UI changes, please attach a screen-shot; otherwise, remove this)

Please review [Ambari Contributing Guide](https://cwiki.apache.org/confluence/display/AMBARI/How+to+Contribute) before opening a pull request.

praveenkjvs committed:
Fix bulk delete api, response code 400
Changes to be committed:
	modified:   ambari-server/src/main/java/org/apache/ambari/server/api/handlers/DeleteHandler.java
	modified:   ambari-server/src/main/java/org/apache/ambari/server/controller/AmbariManagementControllerImpl.java
	modified:   ambari-server/src/main/java/org/apache/ambari/server/controller/internal/DeleteHostComponentStatusMetaData.java
	modified:   ambari-server/src/main/java/org/apache/ambari/server/state/ServiceComponentImpl.java

praveenkjvs commented:
@mpapirkovskyy can you take a look at this pull request?

hapylestat commented:
retest this please

hapylestat commented on ambari-server/src/main/java/org/apache/ambari/server/api/handlers/DeleteHandler.java:
```diff
@@ -83,8 +83,8 @@ protected ResultMetadata convert(RequestStatusMetaData requestStatusMetaData) {
 
-    if (requestStatusMetaData.getClass() != DeleteStatusMetaData.class) {
-      throw new IllegalArgumentException("RequestStatusDetails is not of type DeleteStatusDetails");
+    if (!DeleteStatusMetaData.class.isInstance(requestStatusMetaData)) {
```
please, use the form like `!(requestStatusMetaData instanceof DeleteStatusMetaData)`

praveenkjvs replied:
+1. done thanks

hapylestat commented on ambari-server/src/main/java/org/apache/ambari/server/api/handlers/DeleteHandler.java:
```diff
@@ -83,8 +83,8 @@ protected ResultMetadata convert(RequestStatusMetaData requestStatusMetaData) {
-    if (requestStatusMetaData.getClass() != DeleteStatusMetaData.class) {
-      throw new IllegalArgumentException("RequestStatusDetails is not of type DeleteStatusDetails");
+    if (!DeleteStatusMetaData.class.isInstance(requestStatusMetaData)) {
+      throw new IllegalArgumentException("RequestStatusDetails is not of type DeleteStatusDetails. requestStatusMetaData.getClass= " + requestStatusMetaData.getClass());
```
Exception message is not very much descriptive, let's make it more clear: 
`String.format("Wrong status details class received -  expecting: %s; actual: %s", <class name>, <class name>)`

praveenkjvs replied:
Done. Thanks

hapylestat commented on ambari-server/src/main/java/org/apache/ambari/server/controller/AmbariManagementControllerImpl.java:
```diff
@@ -3605,12 +3605,13 @@ private void checkIfHostComponentsInDeleteFriendlyState(ServiceComponentHostRequ
 
     if (!componentHost.canBeRemoved()) {
-      throw new AmbariException("Host Component cannot be removed"
+      throw new AmbariException("Host Component cannot be removede. Component in undeletable state."
```
Let's replace with something like  `Current host component state prohibiting component removal.`

praveenkjvs replied:
Done. Thanks,

hapylestat commented on ambari-server/src/main/java/org/apache/ambari/server/state/ServiceComponentImpl.java:
```diff
@@ -558,14 +558,15 @@ public void deleteServiceComponentHosts(String hostname, DeleteHostComponentStat
       LOG.info("Deleting servicecomponenthost for cluster" + ", clusterName=" + getClusterName()
           + ", serviceName=" + getServiceName() + ", componentName=" + getName()
-          + ", recoveryEnabled=" + isRecoveryEnabled() + ", hostname=" + sch.getHostName());
+          + ", recoveryEnabled=" + isRecoveryEnabled() + ", hostname=" + sch.getHostName() + ", cuurentState=" + sch.getState());
```
`cuurentState` => `state`

hapylestat commented on ambari-server/src/main/java/org/apache/ambari/server/state/ServiceComponentImpl.java:
```diff
@@ -558,14 +558,15 @@ public void deleteServiceComponentHosts(String hostname, DeleteHostComponentStat
             + ", recoveryEnabled=" + isRecoveryEnabled()
-            + ", hostname=" + sch.getHostName());
+            + ", hostname=" + sch.getHostName()
+            + ", currentState=" + sch.getState());
```
While the components have several states like desired and current, the current status is something what anyone will see by default and is used as generic `state`

`currentState` => `state`

praveenkjvs replied:
Done. Thanks,

hapylestat commented on ambari-server/src/main/java/org/apache/ambari/server/controller/AmbariManagementControllerImpl.java:
```diff
@@ -3605,12 +3605,13 @@ private void checkIfHostComponentsInDeleteFriendlyState(ServiceComponentHostRequ
               + ", hostname=" + request.getHostname()
-              + ", request=" + request);
+              + ", request=" + request
+              + ", currentState=" + componentHost.getState());
```
`currentState` => `state`

praveenkjvs replied:
Done.. Thanks,

hapylestat commented:
@praveenkjvs please make a PR and changes against `branch-2.7` as well, thanks

praveenkjvs committed:
Address review comments and fix test case
 Changes to be committed:
	modified:   ambari-admin/src/main/resources/ui/admin-web/bower.json
	modified:   ambari-server/src/main/java/org/apache/ambari/server/api/handlers/DeleteHandler.java
	modified:   ambari-server/src/main/java/org/apache/ambari/server/controller/AmbariManagementControllerImpl.java
	modified:   ambari-server/src/main/java/org/apache/ambari/server/state/ServiceComponentImpl.java
	modified:   ambari-server/src/test/java/org/apache/ambari/server/controller/AmbariManagementControllerTest.java

hapylestat commented on ambari-admin/src/main/resources/ui/admin-web/bower.json:
```diff
@@ -19,8 +19,5 @@
     "chai": "1.8.0",
     "mocha": "1.14.0",
     "sinon": "1.10.3"
-  },
```
not sure, why this file were modified in scope of current task

praveenkjvs replied:
that was by mistake. My bad.

Thanks for taking this up.
